pkg/controllers/node/node_controller.go

```diff
@@ -0,0 +1,294 @@
+	"github.com/projectcalico/kube-controllers/pkg/controllers/controller"
+	"github.com/projectcalico/libcalico-go/lib/api"
+	"github.com/projectcalico/libcalico-go/lib/client"
+	"github.com/projectcalico/libcalico-go/lib/errors"
+
+	log "github.com/sirupsen/logrus"
+	"k8s.io/apimachinery/pkg/fields"
+	uruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/pkg/api/v1"
+	corecache "k8s.io/client-go/tools/cache"
+
+	"fmt"
+)
+
+// This cache maps a calicoNodeName to its kubernetesNodeName.
+type cache struct {
+	sync.RWMutex
+	nodes map[string]nodeData
+}
+
+type nodeData struct {
+	Name string
+}
+
+// NodeController implements the Controller interface.  It is responsible for monitoring
+// kubernetes nodes and responding to delete events by removing them from the Calico datastore.
+// It keeps a cache of known calico nodes and their corresponding kubernetes nodes to
+// accomplish this.
+type NodeController struct {
+	informer                corecache.Controller
+	k8sResourceCache calicocache.ResourceCache
+	nodeLookupCache   *cache
+	calicoClient            *client.Client
+	k8sClientset            *kubernetes.Clientset
+}
+
+
+// NewNodeController Constructor for NodeController
+func NewNodeController(k8sClientset *kubernetes.Clientset, calicoClient *client.Client) controller.Controller {
+	cacheArgs := calicocache.ResourceCacheArgs{
+		ObjectType: reflect.TypeOf(""),
+		ListFunc:   func() (map[string]interface{}, error) {
+			// Get all nodes from the Calico datastore
+			calicoNodes, err := calicoClient.Nodes().List(api.NodeMetadata{})
+			if err != nil {
+				return nil, err
+			}
+
+			// Iterate through and store the k8s nodes in our cache.
+			m := make(map[string]interface{})
+			for _, calicoNode := range calicoNodes.Items {
+				// find its kubernetes orchRef
+				k8sNodeName := getK8sNodeName(calicoNode)
+				m[k8sNodeName] = calicoNode.Metadata.Name
+			}
+
+			log.Debugf("Found %d nodes in Calico datastore:", len(m))
+			return m, nil
+		},
+		ReconcilerConfig: calicocache.ReconcilerConfig{
+			DisableMissingInDatastore: true,
+			DisableMissingInCache: false,
+			DisableUpdateOnChange: false,
+		},
+	}
+
+	k8sResourceCache := calicocache.NewResourceCache(cacheArgs)
+	nodeLookupCache := cache{nodes: make(map[string]nodeData)}
+
+	// Create a Node watcher.
+	listWatcher := corecache.NewListWatchFromClient(k8sClientset.Core().RESTClient(), "nodes", "", fields.Everything())
+
+	// Bind the Calico cache to kubernetes cache with the help of an informer. This way we make sure that
+	// whenever the kubernetes cache is updated, changes get reflected in the Calico cache as well.
+	_, informer := corecache.NewIndexerInformer(listWatcher, &v1.Node{}, 0, corecache.ResourceEventHandlerFuncs{
+		DeleteFunc: func(obj interface{}) {
+			nodeName, err := extractNodeName(obj)
+			if err != nil {
+				log.WithError(err).Errorf("Error while converting %#v to k8s node", nodeName)
```

Comment:
AFAICT, whenever `extractNodeName` returns an error, the `nodeName` is an empty string, so this message shouldn't include it.

Also, it means we should return early here so we don't call `Delete("")`

Reply:
We can still log out the `obj` we received.